pkg/csilvm/metrics_test.go

```diff
@@ -0,0 +1,327 @@
+	}
+}
+
+func TestReportStorageMetrics(t *testing.T) {
+	// We set an empty prefix as it adds noise to the metric names.
+	const prefix = ""
+	scope := tally.NewTestScope(prefix, nil)
+
+	vgname := testvgname()
+	pvname, pvclean := testpv()
+	defer pvclean()
+	client, clean := startTest(vgname, []string{pvname}, Metrics(scope))
+	defer clean()
+
+	type expect struct {
+		volumes int
+		total   int
+		free    int
+		used    int
+	}
+	check := func(snap tally.Snapshot, exp expect) {
+		gauges := gaugeMap(snap.Gauges())
+		volumes := int(gauges.mustGet("volumes").Value())
+		if volumes != exp.volumes {
+			t.Fatalf("expected %d but got %d", exp.volumes, volumes)
+		}
+		total := int(gauges.mustGet("bytes-total").Value())
+		if total != exp.total {
+			t.Fatalf("expected %d but got %d", exp.total, total)
+		}
+		free := int(gauges.mustGet("bytes-free").Value())
+		if free != exp.free {
+			t.Fatalf("expected %d but got %d", exp.free, free)
+		}
+		used := int(gauges.mustGet("bytes-used").Value())
+		if used != exp.used {
+			t.Fatalf("expected %d but got %d", exp.used, used)
+		}
+		if exp.free+exp.used != exp.total {
+			t.Fatalf("expected %d but got %d", exp.free+exp.used, exp.total)
+		}
+	}
+
+	// Check storage metrics before a volume was created.
+	check(scope.Snapshot(), expect{
+		volumes: 0,
+		total:   100663296,
+		free:    100663296,
+		used:    0,
+	})
+
+	// A single request that fails
+	createVolumeReq := testCreateVolumeRequest()
+	_, err := client.CreateVolume(context.Background(), createVolumeReq)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Check storage metrics after a volume was created.
+	check(scope.Snapshot(), expect{
+		volumes: 1,
+		total:   100663296,
+		free:    16777216,
+		used:    83886080,
+	})
+}
+
+type getOpts struct {
+	tags map[string]string
+}
+
+type getOpt func(*getOpts)
+
+func filterMetricsTags(tags map[string]string) getOpt {
+	return func(opts *getOpts) {
+		if tags == nil {
+			return
+		}
+		opts.tags = tags
+	}
+}
+
+type timerMap map[string]tally.TimerSnapshot
+
+// get finds the TimerSnapshot by name as the default map's key encodes
+// service prefix and tags, too, which makes it inconvenient for looking up
+// timers.
+func (m timerMap) get(name string, opts ...getOpt) (tally.TimerSnapshot, bool) {
+	defaultOpts := getOpts{
+		tags: make(map[string]string),
+	}
+	for _, opt := range opts {
+		if opt != nil {
+			opt(&defaultOpts)
+		}
+	}
+timersLoop:
+	for _, timerSnapshot := range m {
+		if timerSnapshot.Name() == name {
+			// Check that all tags being filtered on are present on the timer.
+			for fkey, fval := range defaultOpts.tags {
+				val, ok := timerSnapshot.Tags()[fkey]
+				if !ok || fval != val {
+					continue timersLoop
+				}
+			}
+			return timerSnapshot, true
+		}
+	}
+	return nil, false
+}
+
+func (m timerMap) mustGet(name string, opts ...getOpt) tally.TimerSnapshot {
+	timer, ok := m.get(name, opts...)
+	if !ok {
+		panic(fmt.Sprintf("cannot find timer %q", name))
```

Comment:
if `mustGet` isn't ever invoked in a parallel goroutine, prefer `t.Fatal` to `panic`

Reply:
I didn't want to push `*testing.T` into these types, but I guess there's no getting away from it.

Reply:
`*testing.T` is the `context` of the `_test.go` world.